lib/util/itinerary.js

```diff
@@ -245,31 +245,133 @@ export function getLegBounds (leg) {
+    if (!aVal && !bVal) return 0
+    // a is not a comparable string, b gets priority
+    if (!aVal) return 1
+    // b is not a comparable string, a gets priority
+    if (!bVal) return -1
+    // a and b are comparable strings, return the sort value
+    if (aVal < bVal) return -1
+    if (aVal > bVal) return 1
+    return 0
+  }
+}
+
+/**
+ * OpenTripPlanner sets the routeSortOrder to -999 by default. So, if that value
+ * is encountered, assume that it actually means that the routeSortOrder is not
+ * set in the GTFS.
+ *
+ * See https://github.com/opentripplanner/OpenTripPlanner/issues/2938
+ * Also see https://github.com/opentripplanner/otp-react-redux/issues/122
+ */
+function getRouteSortOrderValue (val) {
+  return val === -999 ? undefined : val
+}
+
+/**
+ * Create a multi-criteria sort comparator function composed of other sort
+ * comparator functions. Each comparator function will be ran in the order given
+ * until a non-zero comparison value is obtained which is then immediately
+ * returned. If all comparison functions return equivalance, then the values
+ * are assumed to be equivalent.
+ */
+function makeMultiCriteriaSort (...criteria) {
+  return (a, b) => {
+    for (let i = 0; i < criteria.length; i++) {
+      const curCriteriaComparatorValue = criteria[i](a, b)
+      // if the comparison objects are not equivalent, return the value obtained
+      // in this current criteria comparison
+      if (curCriteriaComparatorValue !== 0) {
+        return curCriteriaComparatorValue
+      }
+    }
+    return 0
   }
-  if (aComp < bComp) return -1
-  if (aComp > bComp) return 1
-  return 0
 }
 
+/**
+ * Compares routes for the purposes of sorting and displaying in a user
+ * interface. Due to GTFS feeds having varying levels of data quality, a
+ * multi-criteria sort is needed to account for various differences. The sort
+ * occurs as follows:
+ * 1. Attempt to sort based off of sortOrder. Routes that do not have a valid
+ *    sortOrder will be placed beneath those that do.
+ * 2. In case of the same sortOrder criteria, attempt to sort based off of
+ *    shortNames that are parseable as integers. shortNames that cannot be
+ *    parsed as integers will be placed beneath those that are valid.
+ * 3. In case of the same integer shortName criteria, attempt to sort based
+ *    off of the string shortName. Routes without shortNames will be placed
+ *    beneath those with shortNames.
+ * 4. In case of the same string shortName, attempt to sort based off of
+ *    the string longName.
+ *
+ * @param  {Object} a Route A
+ * @param  {Object} b Route B
+ * @return {number}   A numeric sort comparison value
+ */
+export const routeComparator = makeMultiCriteriaSort(
+  makeNumericValueComparator(obj => getRouteSortOrderValue(obj.sortOrder)),
+  makeNumericValueComparator(obj => parseInt(obj.shortName)),
+  makeObjectStringValueComparator('shortName'),
+  makeObjectStringValueComparator('longName')
```

Comment:
I'm wondering if the `makeObjectStringValueComparator` should just be `makeStringValueComparator` and accept an object getter function like `makeNumericValueComparator`. Seems useful to have parallel structures for these make comparator functions (both for legibility and maintainability).